src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1008,6 +1071,49 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
             )
         }
     }
+
+    fn check_activations(&mut self,
+                         location: Location,
+                         span: Span,
+                         flow_state: &Flows<'cx, 'gcx, 'tcx>)
+    {
+        if !self.tcx.sess.opts.debugging_opts.two_phase_borrows {
+            return;
+        }
+
+        // Two-phase borrow support: For each activation that is newly
+        // generated at this statement, check if it interferes with
+        // another borrow.
+        let domain = flow_state.borrows.operator();
+        let data = domain.borrows();
+        flow_state.borrows.each_gen_bit(|gen| {
+            if gen.is_activation() && // must be activation,
+                !flow_state.borrows.contains(&gen) // and newly generated.
```

Comment:
Won't this be a problem if the activation occurs in a loop?

```rust
let mut y = 0;
let mut z = None;
let mut p = &mut z;
loop {
    z = Some(&mut p); // activate `p`
    **z.as_mut().unwrap() += 1;
    thread::spawn(|| println!("{}", y));
}
```

Reply:
yeah okay. I think I will stop trying to detect that the activation is newly-generated... I cannot remember if my concern was about duplicate errors being reported (which is something we can work around via suppression techniques like the ones we already have in place) or if there was something more fundamental I was concerned about.

Reply:
@arielb1 did you mean to write `let mut p = &mut y;` ?

Reply:
@pnkfelix 

I suppose

Reply:
In any case I have not been able to actually construct a test case demonstrating the concerning situation, namely one where there is a reservation followed by an activation in a loop.  

I'm going to explore removing the requirement that the activation be newly-generated. Niko pointed out that this may lead to problems (namely with a use of `a` conflicting with an earlier use of `a`) but maybe there are simple ways to cope that will be more robust to the kind of scenario @arielb1 has outlined.